### PR TITLE
Add support for NamedTuple streaming using NamedTuples.jl

### PR DESCRIPTION
**Pull request by iamed2**
Passes tests on 0.6, let's see what happens on 0.7.

Also for 0.6 compat but not for NamedTuples:
https://github.com/JuliaData/DataStreams.jl/compare/master...invenia:namedtuples-0.6?expand=1#diff-fce720c43af3c52c862fd7451c7374b8R111
https://github.com/JuliaData/DataStreams.jl/compare/master...invenia:namedtuples-0.6?expand=1#diff-fce720c43af3c52c862fd7451c7374b8R8

### CODE REVIEW

**Comment by iamed2** [comment]
Does Travis's nightly not have named tuples yet? It's still loading NamedTuples.jl despite that package load being behind a conditional.

**Comment by iamed2** [comment]
Fixing some issues locally. Looks like the existing code breaks on Julia `master`.

**Comment by iamed2** [comment]
Okay I think this is ready.

File: REQUIRE [hunk 1]
```diff
@@ -1,3 +1,4 @@
 julia 0.6
 WeakRefStrings 0.3.0
-Nulls 0.0.5
\ No newline at end of file
+Nulls 0.0.5
+NamedTuples 0.4
```
Comment (quinnj) [inline REQUIRE]: I think this is supposed to be 4.0 if I'm reading their releases + METADATA correctly

File: src/DataStreams.jl [hunk 2]
```diff
@@ -601,25 +604,41 @@ function Data.schema(df::NamedTuple{names, T}) where {names, T}
...
+using NamedTuples
+
+# Constraint relaxed for compatability; NamedTuples.NamedTuple does not have parameters
+const Table = NamedTuple
+
+# NamedTuple Data.Source implementation
+# compute Data.Schema on the fly
+function Data.schema(df::NamedTuple)
+    return Data.Schema(Type[eltype(A) for A in values(df)],
+                        collect(map(string, keys(df))), length(df) == 0 ? 0 : length(getfield(df, 1)))
```
Comment (quinnj) [inline src/DataStreams.jl]: Ah, I didn't realize NamedTuples.jl defined `keys` and `values`; that's handy.

File: src/DataStreams.jl [hunk 3]
```diff
@@ -671,7 +690,12 @@ function NamedTuple(sch::Data.Schema{R}, ::Type{S}=Data.Field,
         # for Data.Column or unknown # of rows in Data.Field, we only ever append!, so just allocate empty columns
         rows = ifelse(S == Data.Column, 0, ifelse(!R, 0, sch.rows))
         names = makeunique(Data.header(sch))
-        sink = Base.namedtuple(NamedTuple{names}, (allocate(types[i], rows, reference) for i = 1:length(types))...)
+
+        sink = @static if isdefined(Core, :NamedTuple)
+                Base.namedtuple(NamedTuple{names}, (allocate(types[i], rows, reference) for i = 1:length(types))...)
```
Comment (quinnj) [inline src/DataStreams.jl]: For 0.7, this should just be `NamedTuple{names}((allocate(types[i], rows, reference) for i = 1:length(types)))` (I think that should work)

**Comment by quinnj** [comment]
Thanks for this @iamed2! I pushed one last fix for the NamedTuple constructor that I failed to point out earlier (I gave you the wrong code), but this looks great!

**Comment by iamed2** [comment]
:)

I noticed there was a test missing for some code so I'm going to see if I can bump the coverage up in another PR

**Pull request merged**